### PR TITLE
Auto formatting via `pre-commit.ci`

### PR DESCRIPTION
**Pull request by PointKernel**
This PR uses [`pre-commit.ci`](https://pre-commit.ci/) to enable automatic style fixing for commits in a PR. Users should allow edits by maintainers to get auto-formatting to work.

### CODE REVIEW

**Comment by PointKernel** [comment]
pre-commit.ci run

**Comment by PointKernel** [comment]
pre-commit.ci run

File: .pre-commit-config.yaml
Comment (bdice) [inline .pre-commit-config.yaml:17]: Is there a reason to manually specify the supported file extensions rather than use the type identifiers?

File: .pre-commit-config.yaml
Comment (bdice) [inline .pre-commit-config.yaml]: FYI, this pinning will probably be updated by the “autoupdate” functionality of pre-commit.ci.

**Review by bdice — COMMENTED** [review]
Did this trigger the CI bot correctly? I don’t see any commits from it.

File: README.md
Comment (jrhemstad) [inline README.md]: I think we should leave the old description and just say something like: "By default cuCo will automatically format the code in a PR. If you prefer to do so manually..."

File: .pre-commit-config.yaml
Comment (PointKernel) [reply .pre-commit-config.yaml]: Right, this is why I set the update frequency to `quarterly`. And according to https://github.com/pre-commit-ci/issues/issues/83, there is no plan to provide a schedule-off option.

File: .pre-commit-config.yaml
Comment (PointKernel) [reply .pre-commit-config.yaml:17]: Good question. The type identifier can identify all extensions but `.inl`:
```bash
(base) yunsongw@yunsongw-dt:~/Work/cuCollections/include/cuco/detail$ identify-cli static_map.inl 
["file", "non-executable", "text"]
```
Maybe I should create an issue at `pre-commit` repo to add `.inl` as `header` or `c++`.

File: .pre-commit-config.yaml
Comment (bdice) [reply .pre-commit-config.yaml]: I support `quarterly` just to keep the number of PRs down -- but it does seem like we could update the pinnings before we merge this PR, since it's going to keep clang-format updated anyway. (Unless we're trying to stay pinned to `v11.1.0` for some reason? I don't think the choice of version matters if it's automatically enforced.)

Comment (PointKernel) [reply .pre-commit-config.yaml]: Done

File: .pre-commit-config.yaml
Comment (vyasr) [reply .pre-commit-config.yaml:17]: I think you'd want inl to be considered C++.

File: .pre-commit-config.yaml
Comment (vyasr) [reply .pre-commit-config.yaml]: Do we want to keep this version in line with RAPIDS? AFAIK most cuCollections developers (at least at present) are RAPIDS devs, and having this misaligned could lead to issues if users choose to invoke clang-format via the CLI and they have a different version installed than the one specified by pre-commit. I would be fine if the answer is no, but we should be aware of the issues this discrepancy could cause.

Comment (bdice) [reply .pre-commit-config.yaml]: > I don't think the choice of version matters if it's automatically enforced.

That’s my point: any issues caused by misaligned versions will be auto-fixed by CI, so it doesn’t matter what devs have locally. Moreover, you can’t *really* pin the version because pre-commit CI will open a PR to update it to the latest version every quarter. Personally, I am perfectly okay with keeping the pinnings updated on a quarterly basis and ignoring whatever devs have locally in favor of the pre-commit autofixes.

If we implemented our own “auto-fix” solution in gpuCI similar to pre-commit.ci, then we’d have greater control over version pinnings. However, I don’t think that is needed.

Comment (PointKernel) [reply .pre-commit-config.yaml]: I agree with @bdice, any version mismatch issues will be auto fixed by CI anyway. By all means, it's clearly stated in README that directly using `clang-format` on the command line is not recommended.

Comment (vyasr) [reply .pre-commit-config.yaml]: To be clear, I understood what @bdice was saying about _how_ things would work. I'm asking for a value judgment on whether we think this is a problem. We explicitly declined to move in this direction for RAPIDS because it would prevent using the format-on-write hooks in VSCode, which can't use the pre-commit hook. I don't personally use VSCode, so I don't care about that feature, but before we commit to this approach I just wanted to make sure that everyone involved recognized the limitations since lack of VSCode integration was deemed sufficient reason _not_ to rely on pre-commit versioning in RAPIDS. I'm happy if we just stick with pre-commit though :)

Comment (PointKernel) [reply .pre-commit-config.yaml]: I see your point. This could be a problem not only for VSCode but for all IDEs/editors having auto-formatting support. Technically, users can still stick with the IDE/eidtor `clang-format` hook. It's just any mismatches will be then auto-fixed by CI (or `pre-commit` hook).

Comment (bdice) [reply .pre-commit-config.yaml]: > It's just any mismatches will be then auto-fixed by CI (or `pre-commit` hook).

I'm strongly in favor of trying this approach here. If it works well, I want to eventually revisit that conversation for RAPIDS or cuDF in particular. I don't expect it to be painful, and I suspect it will lead to greater productivity than our current approach of waiting for CI to fail and then identifying/fixing/committing/pushing "style fixes." The current manual approach costs us so much in CI retry time and hinders our iteration speed. Lately I've been doing a lot of work via github.dev (to make small changes, since it's vastly quicker than dealing with local branches / etc. and I can work in lots of browser tabs at once) and automatic CI fixes for style would be so awesome to add to that workflow.

File: .pre-commit-config.yaml
Comment (PointKernel) [reply .pre-commit-config.yaml:17]: Done in https://github.com/pre-commit/identify/pull/273

**Review by jrhemstad — APPROVED** [review]

**Pull request merged**
